File: frontend/src/services/jetbrainsApi.ts
/**
 * JetBrains IDE 集成 API
 *
 * 负责：
 * 1. 检测是否在 IDE 环境中运行
 * 2. 检测后端是否支持 JetBrains 集成
 * 3. 提供 IDE 操作 API（打开文件、显示 Diff 等）
 *
 * 通信方式：
 * - HTTP: 能力检测（/api/jetbrains/capabilities）
 * - RSocket: 所有 IDE 操作和双向通信（/jetbrains-rsocket）
 */

import { resolveServerHttpUrl } from '@/utils/serverUrl'
import { jetbrainsRSocket } from './jetbrainsRSocket'

// ========== 类型定义 ==========

export interface JetBrainsCapabilities {
  supported: boolean
  version: string
}

export interface OpenFileRequest {
  filePath: string
  line?: number
  column?: number
  startOffset?: number
  endOffset?: number
}

export interface ShowDiffRequest {
  filePath: string
  oldContent: string
  newContent: string
  title?: string
}

export interface EditOperation {
  oldString: string
  newString: string
  replaceAll?: boolean
}

export interface ShowMultiEditDiffRequest {
  filePath: string
  edits: EditOperation[]
  currentContent?: string
}

/** 编辑预览请求（权限请求时使用） */
export interface ShowEditPreviewRequest {
  filePath: string
  edits: EditOperation[]
  title?: string
}

/** Markdown 显示请求（计划预览） */
export interface ShowMarkdownRequest {
  content: string
  title?: string
}

// ========== 环境检测 ==========

/**
 * 检测是否在 IDE 环境中运行
 * IDE 环境特征：window.__IDEA_MODE__ = true（由后端注入）
 */
export function isIdeEnvironment(): boolean {
  if (typeof window === 'undefined') return false
  return (window as any).__IDEA_MODE__ === true
}

/**
 * 检测后端是否支持 JetBrains 集成
 */
export async function checkJetBrainsCapabilities(): Promise<JetBrainsCapabilities> {
  try {
    const baseUrl = resolveServerHttpUrl()
    const response = await fetch(`${baseUrl}/api/jetbrains/capabilities`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })

    if (!response.ok) {
      console.warn('[JetBrainsApi] Capabilities check failed:', response.status)
      return { supported: false, version: '' }
    }

    const data = await response.json()
    console.log('[JetBrainsApi] Capabilities:', data)
    return {
      supported: data.supported ?? false,
      version: data.version ?? ''
    }
  } catch (error) {
    console.warn('[JetBrainsApi] Failed to check capabilities:', error)
    return { supported: false, version: '' }
  }
}

/**
 * 检测是否应该启用 JetBrains 集成
 * 条件：IDE 环境 + 后端支持
 */
export async function shouldEnableJetBrainsIntegration(): Promise<boolean> {
  if (!isIdeEnvironment()) {
    console.log('[JetBrainsApi] Not in IDE environment')
    return false
  }

  const capabilities = await checkJetBrainsCapabilities()
  if (!capabilities.supported) {
    console.log('[JetBrainsApi] Backend does not support JetBrains integration')
    return false
  }

  console.log('[JetBrainsApi] JetBrains integration enabled')
  return true
}

// ========== IDE 操作 API ==========

/**
 * JetBrains 桥接服务类
 * 使用 RSocket + Protobuf 与后端 IDE 集成通信
 */
class JetBrainsBridgeService {
  private enabled: boolean = false

  /**
   * 初始化桥接服务
   * 1. 检测环境和后端能力（HTTP）
   * 2. 建立 RSocket 连接
   */
  async init(): Promise<boolean> {
    // 先检测是否应该启用
    const shouldEnable = await shouldEnableJetBrainsIntegration()
    if (!shouldEnable) {
      return false
    }

    // 建立 RSocket 连接
    this.enabled = await jetbrainsRSocket.connect()
    if (this.enabled) {
      console.log('[JetBrainsBridge] Initialized with RSocket')
    }
    return this.enabled
  }

  /**
   * 检查是否已启用
   */
  isEnabled(): boolean {
    return this.enabled && jetbrainsRSocket.isConnected()
  }

  /**
   * 打开文件（RSocket）
   */
  async openFile(request: OpenFileRequest): Promise<boolean> {
    if (!this.enabled) return false
    return jetbrainsRSocket.openFile(request)
  }

  /**
   * 显示 Diff（RSocket）
   */
  async showDiff(request: ShowDiffRequest): Promise<boolean> {
    if (!this.enabled) return false
    return jetbrainsRSocket.showDiff(request)
  }

  /**
   * 显示多编辑 Diff（RSocket）
   */
  async showMultiEditDiff(request: ShowMultiEditDiffRequest): Promise<boolean> {
    if (!this.enabled) return false
    return jetbrainsRSocket.showMultiEditDiff(request)
  }

  /**
   * 显示编辑预览 Diff（权限请求时使用）
   */
  async showEditPreviewDiff(request: ShowEditPreviewRequest): Promise<boolean> {
    if (!this.enabled) return false
    return jetbrainsRSocket.showEditPreviewDiff(request)
  }

  /**
   * 显示 Markdown 内容（计划预览）
   */
  async showMarkdown(request: ShowMarkdownRequest): Promise<boolean> {
    if (!this.enabled) return false
    return jetbrainsRSocket.showMarkdown(request)
  }

  /**
   * 获取项目路径（RSocket）
   */
  async getProjectPath(): Promise<string | null> {
    if (!this.enabled) return null
    return jetbrainsRSocket.getProjectPath()
  }

  /**
   * 获取 IDE 主题（RSocket）
   */
  async getTheme(): Promise<any | null> {
    if (!this.enabled) return null
    return jetbrainsRSocket.getTheme()
  }

  /**
   * 获取语言设置（RSocket）
   */
  async getLocale(): Promise<string> {
    if (!this.enabled) return 'en-US'
    return jetbrainsRSocket.getLocale()
  }

  /**
   * 设置语言（RSocket）
   */
  async setLocale(locale: string): Promise<boolean> {
    if (!this.enabled) return false
    return jetbrainsRSocket.setLocale(locale)
  }

  /**
   * 添加主题变化监听器
   */
  onThemeChange(handler: (theme: any) => void): () => void {
    return jetbrainsRSocket.onThemeChange(handler)
  }

  /**
   * 添加会话命令监听器
   */
  onSessionCommand(handler: (command: any) => void): () => void {
    return jetbrainsRSocket.onSessionCommand(handler)
  }
}

// ========== 单例导出 ==========

export const jetbrainsBridge = new JetBrainsBridgeService()

/**
 * 初始化 JetBrains 集成
 * 应该在应用启动时调用
 */
export async function initJetBrainsIntegration(): Promise<boolean> {
  return jetbrainsBridge.init()
}

File: ai-agent-rpc-api/src/main/kotlin/com/asakii/rpc/api/JetBrainsApi.kt
package com.asakii.rpc.api

import kotlinx.serialization.Serializable

/**
 * JetBrains IDE 集成 API 接口
 *
 * 使用组合模式按功能分组：
 * - jetbrainsApi.capabilities.isSupported()
 * - jetbrainsApi.file.openFile(...)
 * - jetbrainsApi.theme.get()
 * - jetbrainsApi.session.getState()
 * - jetbrainsApi.locale.get()
 */
interface JetBrainsApi {
    val capabilities: JetBrainsCapabilitiesApi
    val file: JetBrainsFileApi
    val theme: JetBrainsThemeApi
    val session: JetBrainsSessionApi
    val locale: JetBrainsLocaleApi
    val project: JetBrainsProjectApi
}

// ========== 能力检测 API ==========

interface JetBrainsCapabilitiesApi {
    fun isSupported(): Boolean
    fun get(): JetBrainsCapabilities
}

// ========== 文件操作 API ==========

interface JetBrainsFileApi {
    fun openFile(request: JetBrainsOpenFileRequest): Result<Unit>
    fun showDiff(request: JetBrainsShowDiffRequest): Result<Unit>
    fun showMultiEditDiff(request: JetBrainsShowMultiEditDiffRequest): Result<Unit>

    /** 展示编辑预览 diff（从文件读取当前内容，应用编辑后显示 diff） */
    fun showEditPreviewDiff(request: JetBrainsShowEditPreviewRequest): Result<Unit>

    /** 展示 Markdown 内容（用于计划预览） */
    fun showMarkdown(request: JetBrainsShowMarkdownRequest): Result<Unit>
}

// ========== 主题 API ==========

interface JetBrainsThemeApi {
    fun get(): JetBrainsIdeTheme?
    fun addChangeListener(listener: (JetBrainsIdeTheme) -> Unit): () -> Unit
}

// ========== 会话管理 API ==========

interface JetBrainsSessionApi {
    /** 接收前端上报的会话状态 */
    fun receiveState(state: JetBrainsSessionState)

    /** 获取当前会话状态 */
    fun getState(): JetBrainsSessionState?

    /** 注册会话状态监听器 */
    fun addStateListener(listener: (JetBrainsSessionState) -> Unit): () -> Unit

    /** 发送会话命令到前端 */
    fun sendCommand(command: JetBrainsSessionCommand)

    /** 注册会话命令监听器（前端使用） */
    fun addCommandListener(listener: (JetBrainsSessionCommand) -> Unit): () -> Unit
}

// ========== 语言设置 API ==========

interface JetBrainsLocaleApi {
    fun get(): String
    fun set(locale: String): Result<Unit>
}

// ========== 项目信息 API ==========

interface JetBrainsProjectApi {
    fun getPath(): String
}

// ========== 默认实现（不支持 JetBrains 集成时使用）==========

object DefaultJetBrainsApi : JetBrainsApi {
    override val capabilities = object : JetBrainsCapabilitiesApi {
        override fun isSupported() = false
        override fun get() = JetBrainsCapabilities(supported = false)
    }

    override val file = object : JetBrainsFileApi {
        private val error = UnsupportedOperationException("JetBrains integration not available")
        override fun openFile(request: JetBrainsOpenFileRequest) = Result.failure<Unit>(error)
        override fun showDiff(request: JetBrainsShowDiffRequest) = Result.failure<Unit>(error)
        override fun showMultiEditDiff(request: JetBrainsShowMultiEditDiffRequest) = Result.failure<Unit>(error)
        override fun showEditPreviewDiff(request: JetBrainsShowEditPreviewRequest) = Result.failure<Unit>(error)
        override fun showMarkdown(request: JetBrainsShowMarkdownRequest) = Result.failure<Unit>(error)
    }

    override val theme = object : JetBrainsThemeApi {
        override fun get(): JetBrainsIdeTheme? = null
        override fun addChangeListener(listener: (JetBrainsIdeTheme) -> Unit) = { }
    }

    override val session = object : JetBrainsSessionApi {
        override fun receiveState(state: JetBrainsSessionState) {}
        override fun getState(): JetBrainsSessionState? = null
        override fun addStateListener(listener: (JetBrainsSessionState) -> Unit) = { }
        override fun sendCommand(command: JetBrainsSessionCommand) {}
        override fun addCommandListener(listener: (JetBrainsSessionCommand) -> Unit) = { }
    }

    override val locale = object : JetBrainsLocaleApi {
        override fun get() = "en-US"
        override fun set(locale: String) = Result.failure<Unit>(
            UnsupportedOperationException("JetBrains integration not available")
        )
    }

    override val project = object : JetBrainsProjectApi {
        override fun getPath() = ""
    }
}

// ========== 数据模型 ==========

data class JetBrainsCapabilities(
    val supported: Boolean,
    val version: String = "1.0"
)

data class JetBrainsOpenFileRequest(
    val filePath: String,
    val line: Int? = null,
    val column: Int? = null,
    val startOffset: Int? = null,
    val endOffset: Int? = null
)

data class JetBrainsShowDiffRequest(
    val filePath: String,
    val oldContent: String,
    val newContent: String,
    val title: String? = null
)

data class JetBrainsEditOperation(
    val oldString: String,
    val newString: String,
    val replaceAll: Boolean = false
)

data class JetBrainsShowMultiEditDiffRequest(
    val filePath: String,
    val edits: List<JetBrainsEditOperation>,
    val currentContent: String? = null
)

/** 编辑预览请求（从文件读取当前内容，应用编辑后显示 diff） */
data class JetBrainsShowEditPreviewRequest(
    val filePath: String,
    val edits: List<JetBrainsEditOperation>,
    val title: String? = null
)

/** Markdown 内容显示请求 */
data class JetBrainsShowMarkdownRequest(
    val content: String,
    val title: String? = null
)

@Serializable
data class JetBrainsIdeTheme(
    val background: String,
    val foreground: String,
    val borderColor: String,
    val panelBackground: String,
    val textFieldBackground: String,
    val selectionBackground: String,
    val selectionForeground: String,
    val linkColor: String,
    val errorColor: String,
    val warningColor: String,
    val successColor: String,
    val separatorColor: String,
    val hoverBackground: String,
    val accentColor: String,
    val infoBackground: String,
    val codeBackground: String,
    val secondaryForeground: String,
    val fontFamily: String,
    val fontSize: Int,
    val editorFontFamily: String,
    val editorFontSize: Int
)

data class JetBrainsSessionSummary(
    val id: String,
    val title: String,
    val sessionId: String? = null,
    val isGenerating: Boolean = false,
    val isConnected: Boolean = false,
    val isConnecting: Boolean = false
)

data class JetBrainsSessionState(
    val sessions: List<JetBrainsSessionSummary>,
    val activeSessionId: String? = null
)

enum class JetBrainsSessionCommandType {
    UNSPECIFIED,
    SWITCH,
    CREATE,
    CLOSE,
    RENAME,
    TOGGLE_HISTORY,
    SET_LOCALE,
    DELETE,  // 删除历史会话（直接删除文件，不需要同步）
    RESET    // 重置/清空当前会话（不新建 Tab）
}

data class JetBrainsSessionCommand(
    val type: JetBrainsSessionCommandType,
    val sessionId: String? = null,
    val newName: String? = null,
    val locale: String? = null
)
